File: dongho-kim/2주차/Daily Temperatures/s.java
import java.util.*;

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Deque<Integer> stack = new ArrayDeque<>();

        int[] answer = new int[temperatures.length];
        for (int curr_idx = 0; curr_idx < temperatures.length; curr_idx++) {
            int curr_temp = temperatures[curr_idx];

            while (!stack.isEmpty() && curr_temp > temperatures[stack.peek()]) {
                int prev_idx = stack.pop();
                int dist = curr_idx - prev_idx;
                answer[prev_idx] = dist;
            }

            stack.push(curr_idx);
        }

        return answer;
    }
}

File: dongho-kim/2주차/Longest Valid Parentheses/s.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = []
        matched = [False] * len(s)

        for i, c in enumerate(s):
            if c == '(':
                stack.append(i)
                continue

            if c == ')':
                if len(stack) == 0:
                    continue

                prev_idx = stack.pop()

                matched[prev_idx] = True
                matched[i] = True

        answer = 0
        curr = 0
        for m in matched:
            if m:
                curr += 1
                continue

            answer = max(answer, curr)
            curr = 0
        answer = max(answer, curr)

        return answer
